sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -1538,6 +1585,32 @@ pi_result _pi_platform::populateDeviceCacheIfNeeded() {
           delete[] ZeSubdevices;
           return Result;
         }
+
+        // sub-sub-device
+        // get all the ordinals for the sub-sub-devices
+        pi_uint32 SubSubDevicesCount = 0;
+        std::vector<int> Ordinals;
+        std::vector<ze_command_queue_group_properties_t> AllQueueProperties;
+        Result = getCmdQueueOrdinals(PiSubDevice.get(), SubSubDevicesCount, Ordinals, AllQueueProperties);
+        if (Result != PI_SUCCESS) {
+          return Result;
+        }
+
+        // Create PI sub-sub-devices with the sub-device for all the ordinals
+        for (uint32_t J = 0; J < SubSubDevicesCount; ++J) {
+          // TODO: check if a device can be it's own parent
+          std::unique_ptr<_pi_device> PiSubSubDevice(
+            new _pi_device(ZeSubdevices[I], this, PiSubDevice.get()));
+          pi_result Result = initializeWithOrdinal(PiSubSubDevice.get(), Ordinals[J], AllQueueProperties);
+          if (Result != PI_SUCCESS) {
+            return Result;
+          }
+
+          // save pointers to sub-sub-devices for quick retrieval in the future.
+          PiSubDevice->SubDevices.push_back(PiSubSubDevice.get());
```

Comment:
So, you save the pi_device for sub-sub-devices to just sub-devices. It's not clear (at that code is missing) how you retrieve the right pi_device-s when piDevicePartition is called on a sub-device.

Reply:
my understanding was that, as each sub-device is a `device`, when piDevicePartition is called for a sub-device (aka a device), it's sub-device array will have it's sub-sub-devices. this works because initialize() function treats any sub-device as a device and populates the sub-sub-devices into sub-devices. then we can return it with the same piDevicePartition logic.

Reply:
Good point. I see now, this is "sub-sub" because of `PiSubDevice->SubDevices`.
Anyways, even though it seems to be OK, we need a test to fully verify that.

Reply:
@smaslov-intel, I've uploaded a test here: https://github.com/intel/llvm-test-suite/pull/345. The test passes in local testing.